### PR TITLE
Add GCN

### PR DESCRIPTION
**Pull request by Derek-Wds**
Hi this is the PR about adding GCN to sqlflow, which relates to [#2714](https://github.com/sql-machine-learning/sqlflow/issues/2714), [#78](https://github.com/sql-machine-learning/models/issues/78).

This sql command is used to train the model, which can produce a result of accuracy about 83% on validation dataset.
```sql
SELECT cora.node.id, features, label as class, cora.edge.from_node_id, cora.edge.to_node_id FROM cora.node
LEFT JOIN cora.edge ON (cora.node.id = cora.edge.from_node_id OR cora.node.id = cora.edge.to_node_id)
ORDER BY cora.node.id
TO TRAIN sqlflow_models.GCN
WITH model.nhid=16, model.nclass=7, 
     model.epochs=200, model.train_ratio=0.15, 
     model.eval_ratio=0.2, validation.metrics="CategoricalAccuracy"
COLUMN DENSE(features)
LABEL class
INTO sqlflow_models.gcn_model;
```

If there is any question, please let me know.

### CODE REVIEW

File: sqlflow_models/gcn.py [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+        features = features[idx]
+        labels = labels[idx]
+        # preprocess
+        features = GCN.normalize_feature(features)
+        labels = GCN.encode_onehot(labels)
+        adjacency = sp.coo_matrix((np.ones(len(edges)),
+                    (edges[:, 0], edges[:, 1])),
+                    shape=(features.shape[0], features.shape[0]), dtype="float32")
+
+        adjacency = adjacency + adjacency.T.multiply(adjacency.T > adjacency) - adjacency.multiply(adjacency.T > adjacency)
```
Comment (lhw362950217) [inline sqlflow_models/gcn.py:169]: as for 0,1 matrix, the `- adjacency.multiply(adjacency.T > adjacency)` is not necessary?

File: sqlflow_models/gcn.py [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+    def sqlflow_evaluate_loop(self, x, metric_names):
+        """Customed evaluation, can only support calculating the accuracy."""
+        with open('./results.p', 'rb') as f:
+            results = pickle.load(f)
+            eval_result = results['evaluation']
+        return eval_result
+
+    def sqlflow_predict_one(self, sample):
+        """Customed prediction, sample must be the node id."""
+        with open('./results.p', 'rb') as f:
```
Comment (lhw362950217) [inline sqlflow_models/gcn.py]: we don't need to open the file each time the sqlflow_predict_one is called

File: sqlflow_models/gcn.py (hunk 1 above)
Comment (Derek-Wds) [reply sqlflow_models/gcn.py:169]: From my perspective, this is still necessary and it aims to build the bidirectional edges of the graph.

File: sqlflow_models/gcn.py (hunk 2 above)
Comment (Derek-Wds) [reply sqlflow_models/gcn.py]: Yeah sure! This is indeed inefficient, but how should we retrieve the result from the file during prediction?

File: sqlflow_models/gcn.py [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+import tensorflow.keras.backend as K
+import scipy.sparse as sp
+import numpy as np
+import pickle, copy
+
+
+# Graph Convolutional Neural Networks
+class GCN(tf.keras.Model):
+    def __init__(self, nhid, nclass, epochs, train_ratio, eval_ratio, early_stopping=True, dropout=0.5, nlayer=2, feature_columns=None):
+        """GCN
```
Comment (typhoonzero) [inline sqlflow_models/gcn.py]: Can remove above `# Graph Convolutional Neural Networks` line and add some simple description about GCN in this doc string.

File: sqlflow_models/gcn.py [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+            type nlayer: int.
+        :param feature_columns: a list of tf.feature_column. (Not used in this model)
+            type feature_columns: list.
+        """
+        super(GCN, self).__init__()
+
+        self.gc_layers = list()
+        self.gc_layers.append(GCNLayer(nhid, kernel_regularizer=tf.keras.regularizers.l2(5e-4)))
+        if nlayer > 2:
+            for i in range(nlayer-2):
```
Comment (typhoonzero) [inline sqlflow_models/gcn.py]: Line 37 seems can be moved into this loop?

File: sqlflow_models/gcn.py [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+    def sqlflow_train_loop(self, x):
+        """Customized training function."""
+        # load data
+        ids, ids_check, features, labels, edges, edge_check = list(), dict(), list(), list(), list(), dict()
+        from_node = 0
+        for inputs, label in x:
+            id = inputs['id'].numpy().astype(np.int32)
+            feature = inputs['features'].numpy().astype(np.float32)
+            from_node = inputs['from_node_id'].numpy().astype(np.int32)
+            to_node = inputs['to_node_id'].numpy().astype(np.int32)
```
Comment (typhoonzero) [inline sqlflow_models/gcn.py]: Maybe we can set some variables when init the model class to set the column names used for `id, features, from_node_id, to_node_id` since the column name may be different.

File: sqlflow_models/gcn.py (hunk 2 above)
Comment (typhoonzero) [reply sqlflow_models/gcn.py]: Can we try loading the file in `__init__` if the file exists?

Comment (Derek-Wds) [reply sqlflow_models/gcn.py]: Yes, this is what I'm thinking rn. If sqlflow will init the model once during the evaluation and prediction phase, then we could load the file in the `__init__` and use it directly.

File: sqlflow_models/gcn.py (hunk 5 above)
Comment (Derek-Wds) [reply sqlflow_models/gcn.py]: Yeah sure, this will be more flexible when people create dataset with different column names. I will update this later :)

**Comment by Derek-Wds** [comment]
I have updated the codes following @lhw362950217 and @typhoonzero's suggestions. Thanks!

File: sqlflow_models/gcn.py (hunk 5 above)
Comment (Derek-Wds) [reply sqlflow_models/gcn.py]: One thing I have to mention is that, `id`, `from_node_id` and `to_node_id` should be `integer` since they are treated as the index of the node in a certain order. Otherwise, it may cause some problems when we build the graph with this code. 

I will make this clear in the GCN documentation later.

File: sqlflow_models/gcn.py (hunk 4 above)
Comment (Derek-Wds) [reply sqlflow_models/gcn.py]: The first layer was intended to handle the different type of `features` (either `sparse` or not). I have updated the codes to add this feature.

File: sqlflow_models/gcn.py [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+        # store the results in a pickled file
+        with open('./results.pkl', 'wb') as f:
+            results = dict()
+            for i in range(len(ids)):
+                results[str(ids[i])] = predicted[i]
+            results['evaluation'] = result
+            pickle.dump(results, f)
+            self.results = results
+
+    def sqlflow_evaluate_loop(self, x, metric_names):
```
Comment (sneaxiy) [inline sqlflow_models/gcn.py:295]: If we can only do the evaluation once after training, do we need this function?

File: sqlflow_models/gcn.py [hunk 7]
```diff
@@ -0,0 +1,408 @@
...
+        """Customed evaluation, can only support calculating the accuracy."""
+        assert self.results is not None, "Please make sure to train the model first."
+        eval_result = self.results['evaluation']
+        return eval_result
+
+    def sqlflow_predict_one(self, sample):
```
Comment (sneaxiy) [inline sqlflow_models/gcn.py:301]: Same as `sqlflow_evaluate_loop`.

File: sqlflow_models/gcn.py [hunk 8]
```diff
@@ -0,0 +1,408 @@
...
+    
+    def test(self, mask, return_loss=False):
+        '''Test the results on the model. Return accuracy'''
+        logits = self.predict(data=[self.features, self.adjacency])
+
+        test_mask_logits = tf.gather_nd(logits, tf.where(mask))
+        masked_labels = tf.gather_nd(self.labels, tf.where(mask))
+
+        ll = tf.equal(tf.argmax(masked_labels, -1), tf.argmax(test_mask_logits, -1))
+        accuarcy = tf.reduce_mean(tf.cast(ll, dtype=tf.float32))
```
Comment (sneaxiy) [inline sqlflow_models/gcn.py]: Typo. `accuarcy` -> `accuracy`.

File: sqlflow_models/gcn.py [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+        self.gc_layers = list()
+        self.gc_layers.append(GCNLayer(nhid, kernel_regularizer=tf.keras.regularizers.l2(5e-4), sparse_input=sparse_input))
+        for i in range(nlayer-1):
+            self.gc_layers.append(GCNLayer(nhid, kernel_regularizer=tf.keras.regularizers.l2(5e-4)))
+        self.gc_layers.append(GCNLayer(nclass))
+        self.keep_prob = 1 - dropout
+        self.dropout = tf.keras.layers.Dropout(dropout)
+        self.nshape = None
+        self.train_ratio = train_ratio
+        self.eval_ratio = eval_ratio
```
Comment (sneaxiy) [inline sqlflow_models/gcn.py:60]: Check `0 < train_ratio < 1`, `0 < eval_ratio < 1` and `0 < train_ratio + eval_traio < 1`?

File: sqlflow_models/gcn.py [hunk 10]
```diff
@@ -0,0 +1,408 @@
...
+        random_tensor += tf.random_uniform(noise_shape)
+        dropout_mask = tf.cast(tf.floor(random_tensor), dtype=tf.bool)
+        pre_out = tf.sparse_retain(x, dropout_mask)
+        return pre_out * (1./keep_prob)
+
+    @staticmethod
+    def encode_onehot(labels):
+        classes = set(labels)
+        classes_dict = {c: np.identity(len(classes))[i, :] for i, c in enumerate(classes)}
+        labels_onehot = np.array(list(map(classes_dict.get, labels)), dtype=np.int32)
```
Comment (sneaxiy) [inline sqlflow_models/gcn.py:115]: Simplify by `labels_onehot = np.array(map(classes_dict.get, labels), dtype=np.int32)`?

File: sqlflow_models/gcn.py (hunk 8 above)
Comment (Derek-Wds) [reply sqlflow_models/gcn.py]: Thanks for pointing out, I will fix this later.

File: sqlflow_models/gcn.py (hunk 9 above)
Comment (Derek-Wds) [reply sqlflow_models/gcn.py:60]: Yeah, I will do it later and also take a look if there exist other parameters should be checked as well. Thanks!

File: sqlflow_models/gcn.py (hunk 10 above)
Comment (Derek-Wds) [reply sqlflow_models/gcn.py:115]: Well, this seems to be not working, since when we calculating the `loss` and `accuracy`, we will use `tf.gather_nd` to get masked labels, and tensorflow does not support converting a `numpy array` with `map` value to a `tensor`.

File: sqlflow_models/gcn.py (hunk 6 above)
Comment (Derek-Wds) [reply sqlflow_models/gcn.py:295]: I guess we still need it if we want to have `EVALUATE` and `PREDICT` commands through sqlflow. Otherwise, we have to print out the results for both evaluation and prediction during training phase. (This seems to be a bad idea for me.) What do other people think? @typhoonzero, @lhw362950217 (Ideas are greatly appreciated!)

Comment (lhw362950217) [reply sqlflow_models/gcn.py:295]: I think we may keep it to adapt the `PREDICT` and `EVALUATE` statements.

Comment (typhoonzero) [reply sqlflow_models/gcn.py:295]: I agree with @Derek-Wds .

File: tests/test_gcn.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+        self.model = sqlflow_models.GCN(nhid=16, nclass=4, epochs=20, train_ratio=0.2, eval_ratio=0.15)
+        self.model_class = sqlflow_models.GCN
+
+    def test_train_and_predict(self):
+        self.setUp()
+        self.model.compile(optimizer=optimizer(),
+                           loss='categorical_crossentropy')
+        self.model.sqlflow_train_loop(zip(self.inputs, self.labels))
+        metric = evaluate([self.model.features, self.model.adjacency], self.model.labels, self.model)
+        print(metric)
```
Comment (typhoonzero) [inline tests/test_gcn.py]: Can remove this line for debugging.

**Review by typhoonzero — DISMISSED** [review]
LGTM++

File: tests/test_gcn.py (hunk 11 above)
Comment (Derek-Wds) [reply tests/test_gcn.py]: Sure, I will remove it now. Thanks!

**Review by typhoonzero — APPROVED** [review]
